### PR TITLE
Use `strict=True` with `zip` to ensure length equality

### PR DESCRIPTION
**Pull request by DanielYang59**
### Summary

- Use `strict=True` when `zip` to ensure length equality @janosh, to finish https://github.com/materialsproject/pymatgen/pull/4009#discussion_r1724511180

### TODOs

- [x] Let's see how many tests are failing 
- [x] Make sure `strict` keyword is for `zip` not other function

### CODE REVIEW

File: src/pymatgen/cli/pmg_config.py [hunk 1]
```diff
@@ -272,7 +272,7 @@ def add_config_var(tokens: list[str], backup_suffix: str) -> None:
             print(f"Existing {rc_path} backed up to {rc_path}{backup_suffix}")
         dct = loadfn(rc_path)
     special_vals = {"true": True, "false": False, "none": None, "null": None}
-    for key, val in zip(tokens[::2], tokens[1::2], strict=False):
+    for key, val in zip(tokens[::2], tokens[1::2], strict=True):
```
Comment (DanielYang59) [inline src/pymatgen/cli/pmg_config.py:275]: TODO: Perhaps use itertools for this?

File: src/pymatgen/io/abinit/inputs.py [hunk 2]
```diff
@@ -663,7 +663,7 @@ def pop_vars(self, keys):
         Example:
             inp.pop_vars(["ionmov", "optcell", "ntime", "dilatmx"])
         """
-        return self.remove_vars(keys, strict=False)
+        return self.remove_vars(keys, strict=True)
```
Comment (DanielYang59) [inline src/pymatgen/io/abinit/inputs.py:666]: Revert this, not `zip`

File: src/pymatgen/io/abinit/inputs.py [hunk 3]
```diff
@@ -807,7 +807,7 @@ def __setitem__(self, key, value):
             and hasattr(self, "_vars")
             and any(tol in self._vars and tol != key for tol in _TOLVARS_SCF)
         ):
-            logger.info(f"Replacing previously set tolerance variable: {self.remove_vars(_TOLVARS_SCF, strict=False)}.")
+            logger.info(f"Replacing previously set tolerance variable: {self.remove_vars(_TOLVARS_SCF, strict=True)}.")
```
Comment (DanielYang59) [inline src/pymatgen/io/abinit/inputs.py:810]: Revert this

File: src/pymatgen/io/abinit/inputs.py [hunk 4]
```diff
@@ -987,14 +987,14 @@ def pop_tolerances(self):
         Remove all the tolerance variables present in self.
         Return dictionary with the variables that have been removed.
         """
-        return self.remove_vars(_TOLVARS, strict=False)
+        return self.remove_vars(_TOLVARS, strict=True)
```
Comment (DanielYang59) [inline src/pymatgen/io/abinit/inputs.py:990]: Revert

File: src/pymatgen/io/abinit/inputs.py [hunk 5]
```diff
@@ -987,14 +987,14 @@ def pop_tolerances(self):
...
 
     def pop_irdvars(self):
         """
         Remove all the `ird*` variables present in self.
         Return dictionary with the variables that have been removed.
         """
-        return self.remove_vars(_IRDVARS, strict=False)
+        return self.remove_vars(_IRDVARS, strict=True)
```
Comment (DanielYang59) [inline src/pymatgen/io/abinit/inputs.py:997]: Revert

File: src/pymatgen/io/cp2k/inputs.py [hunk 6]
```diff
@@ -353,7 +353,7 @@ def __delitem__(self, key):
     def __sub__(self, other):
         return self.__delitem__(other)
 
-    def setitem(self, key, value, strict=False):
+    def setitem(self, key, value, strict=True):
```
Comment (DanielYang59) [inline src/pymatgen/io/cp2k/inputs.py:356]: Revert

File: src/pymatgen/io/cp2k/inputs.py [hunk 7]
```diff
@@ -418,7 +418,7 @@ def get_keyword(self, d, default=None):
                 return v
         return default
 
-    def update(self, dct: dict, strict=False) -> Section:
+    def update(self, dct: dict, strict=True) -> Section:
```
Comment (DanielYang59) [inline src/pymatgen/io/cp2k/inputs.py:421]: Revert

File: src/pymatgen/io/cp2k/inputs.py [hunk 8]
```diff
@@ -446,7 +446,7 @@ def update(self, dct: dict, strict=False) -> Section:
         return self
 
     @staticmethod
-    def _update(d1, d2, strict=False):
+    def _update(d1, d2, strict=True):
```
Comment (DanielYang59) [inline src/pymatgen/io/cp2k/inputs.py:449]: Revert

File: tests/core/test_composition.py [hunk 9]
```diff
@@ -31,7 +31,7 @@ def setUp(self):
 
         self.indeterminate_comp = [
             Composition.ranked_compositions_from_indeterminate_formula("Co1", lock_if_strict=True),
-            Composition.ranked_compositions_from_indeterminate_formula("Co1", lock_if_strict=False),
+            Composition.ranked_compositions_from_indeterminate_formula("Co1", lock_if_strict=True),
```
Comment (DanielYang59) [inline tests/core/test_composition.py:34]: revert

File: tests/core/test_composition.py [hunk 10]
```diff
@@ -829,7 +829,7 @@ def test_math(self):
 
         # test get_energy()
         assert pots.get_energy(fe_o2) == approx(5.2)
-        assert fe_pot.get_energy(fe_o2, strict=False) == approx(1)
+        assert fe_pot.get_energy(fe_o2, strict=True) == approx(1)
```
Comment (DanielYang59) [inline tests/core/test_composition.py:832]: revert

File: tests/io/abinit/test_inputs.py [hunk 11]
```diff
@@ -93,7 +93,7 @@ def test_api(self):
         with pytest.raises(KeyError) as exc:
             inp.remove_vars("foo", strict=True)
         assert "key='foo' not in self:" in str(exc.value)
-        assert not inp.remove_vars("foo", strict=False)
+        assert not inp.remove_vars("foo", strict=True)
```
Comment (DanielYang59) [inline tests/io/abinit/test_inputs.py:96]: revert

File: tests/io/aims/test_sets/test_input_set.py [hunk 12]
```diff
@@ -295,7 +295,7 @@ def test_input_set():
     assert check_file(control_in_str, in_set.control_in)
     assert params_json == json.loads(in_set.params_json)
 
-    in_set.remove_parameters(keys=["relax_geometry"], strict=False)
+    in_set.remove_parameters(keys=["relax_geometry"], strict=True)
```
Comment (DanielYang59) [inline tests/io/aims/test_sets/test_input_set.py:298]: revert

**Review by janosh — COMMENTED** [review]
thanks @DanielYang59! that was way faster than i expected. 👍 

did you notice any cases where `strict=True` stood out as potentially problematic? knowing pymatgen's test coverage, i assume there are cases where `strict=True` wouldn't raise a CI warning despite `strict=False` being required for backwards compatibility

**Comment by DanielYang59** [comment]
> did you notice any cases where `strict=True` stood out as potentially problematic? 

Sorry I didn't really go through each of these replacements (just a batch replace and fixed those CI failures in 89dc24775a9926f335ba271f477dd58890aae35d and b149ebb52918a35151a127cd99b555c809473e90). Perhaps I should mark it as draft now and fully go through all changes later.

> i assume there are cases where `strict=True` wouldn't raise a CI warning 

With `strict=True` [a `ValueError` would be raised](https://github.com/materialsproject/pymatgen/actions/runs/10503059600/job/29095707287) directly instead of warning. So it would be a good thing (reveal untested code) and a bad thing (might lead to many breakages from untested code).

**Comment by janosh** [comment]
> With strict=True [a ValueError would be raised](https://github.com/materialsproject/pymatgen/actions/runs/10503059600/job/29095707287) directly instead of warning

sorry, i meant error, not warning

> So it would be a good thing (reveal untested code) and a bad thing (might lead to many breakages from untested code).

agreed

> Perhaps I should mark it as draft now and fully go through all changes later.

not strictly necessary imo. these issues are very hard to spot by just looking at the code. overall less time spent reporting and fixing errors as they occur

**Comment by DanielYang59** [comment]
> not strictly necessary imo. these issues are very hard to spot by just looking at the code. overall less time spent reporting and fixing errors as they occur

Yes that's what I thought... Though sounds a bit irresponsible, but it's quite hard just by eyeballing

**Review by janosh — APPROVED** [review]
thanks as always @DanielYang59!

**Comment by DanielYang59** [comment]
No problem. Thanks for reviewing. Fingers crossed that this PR wouldn't be pinged a hundred times :)

**Pull request merged**